reasoner/planner/CostEstimator.java

```diff
@@ -0,0 +1,384 @@
+            this.costEstimator = costEstimator;
+            this.logicMgr = costEstimator.logicMgr;
+            this.conjunction = conjunction;
+            this.fullAnswerCount = -1;
+        }
+
+        public long estimateAllAnswers() {
+            return this.fullAnswerCount;
+        }
+
+        public long estimateAnswers(Set<Variable> projectToVariables, Set<Resolvable<?>> includedResolvables) {
+            List<LocalEstimate> enabledEstimates = iterate(includedResolvables)
+                    .flatMap(resolvable -> iterate(multivarEstimates.get(resolvable)))
+                    .toList();
+            return computeCostCover(enabledEstimates, projectToVariables);
+        }
+
+        private long computeCostCover(List<LocalEstimate> multivarEstimates, Set<Variable> projectToVariables) {
+            // Does a greedy set cover
+            Map<Variable, CostCover> costCover = new HashMap<>(unaryCostCover);
+            multivarEstimates.sort(Comparator.comparing(x -> x.answerEstimate(this, projectToVariables)));
+            for (LocalEstimate multivarEstimate : multivarEstimates) {
+                Set<Variable> interestingSubsetOfVariables = multivarEstimate.variables.stream()
+                        .filter(projectToVariables::contains).collect(Collectors.toSet());
+
+                long currentCostToCover = CostCover.costToCover(interestingSubsetOfVariables, costCover);
+                if (currentCostToCover > multivarEstimate.answerEstimate(this, interestingSubsetOfVariables)) {
+                    CostCover variablesNowCoveredBy = new CostCover(multivarEstimate.answerEstimate(this, interestingSubsetOfVariables));
+                    interestingSubsetOfVariables.forEach(v -> costCover.put(v, variablesNowCoveredBy));
+                }
+            }
+
+            long ret = CostCover.costToCover(projectToVariables, costCover);
+            assert ret > 0;             // Flag in tests if it happens.
+            return Math.max(ret, 1);    // Don't do stupid stuff in prod when it happens.
+        }
+
+        private void registerTriggeredRules() {
+            resolvables().filter(this::isInferrable).map(Resolvable::asConcludable).forEachRemaining(concludable -> {
+                Map<Rule, Set<Unifier>> unifiers = logicMgr.applicableRules(concludable);
+                iterate(unifiers.keySet()).forEachRemaining(rule -> {
+                    costEstimator.registerConjunction(rule.condition().conjunction());
+                });
+            });
+        }
+
+        private boolean isInferrable(Resolvable<?> resolvable) {
+            return resolvable.isConcludable() &&
+                    !logicMgr.applicableRules(resolvable.asConcludable()).isEmpty();
+        }
+
+        private FunctionalIterator<Resolvable<?>> resolvables() {
+            return iterate(logicMgr.compile(conjunction));
+        }
+
+
+        private void initializeOnce() {
```

Comment:
Collects the "estimates" which are all we need to answer any of the queries of interest